preventing file open and bad defer close call

josedh opened the pull request:
The previous function was calling defer Close() before checking the error. THis could potential lead to a panic if calling close on a nil pointer. Fixed by removing the os.Open call altogether and using os.Stat instead.

Lomanic commented:
Looks like this kind of check (checking if PID exists) is only done on the linux platform. Shall we extend this check to other platforms or remove it here instead?

josedh commented:
I went ahead and added the process check to all system types. Moved away from file check and instead tried to send a 0 signal to the process to check for process existence.

Lomanic commented:
Please test this on Windows at least once (process.Processes() returns an empty list with this PR as an example). See #607 (signals are a POSIX thing, Windows is not POSIX).

Otherwise, please remove your last commit and this will be merged as-is.

josedh commented:
@Lomanic: What if I simply update the return line (631) to just return the new process and nil. It seems the call to os.FindProcess * should * handle the check for existence of the process in windows, and thus be sufficient to cover this case. See: https://golang.org/src/os/exec_windows.go#L90

Lomanic commented:
Please reset your branch to cd515b51f3c5cba146c9988cc1b01c3a9c2c72ca (`git reset --hard 13375e2f9c9b4f2cfca1f851acde9a674b27fce8 && git push --force`) and this will be merged as-is, as I tested this commit working properly on Linux. Your subsequent commits were obviously not tested (it currently doesn't even compile on Windows as proc variable is not used, and you didn't seem to follow links posted in the issue I linked in my previous comment to implement the check properly for POSIX OSes) and my bandwidth is limited, so let's keep this PR short. Thanks in advance.

josedh commented:
Done. Sorry about the mess. I'll try to revisit the rest of the platforms at some point and hopefully add an appropriate fix for those as well

Lomanic commented:
Your last commit was not tested (it did not compile with undefined var p). I fixed it in master.

josedh commented:
Thanks. Seems like my vim-go installation was messed up and it reported "build sucessful" when I ran :GoBuild
